crates/bevy_ecs/src/query/iter.rs

```diff
@@ -1465,6 +2097,183 @@ impl<'w, 's, D: QueryData, F: QueryFilter, I: Iterator<Item: Borrow<Entity>>> De
+            // This means `world` has permission to access the data we use.
+            tables: &world.storages().tables,
+            fetch,
+            entity_iter: entity_list.into_iter(),
+        }
+    }
+
+    /// # Safety
+    /// The lifetime here is not restrictive enough for Fetch with &mut access,
+    /// as calling `fetch_next_aliased_unchecked` multiple times can produce multiple
+    /// references to the same component, leading to unique reference aliasing.
+    ///
+    /// It is always safe for shared access.
+    /// `entity` must stem from `self.entity_iter`, and not have been passed before.
+    #[inline(always)]
+    unsafe fn fetch_next_aliased_unchecked(&mut self, entity: Entity) -> D::Item<'w> {
+        let (location, archetype, table);
+        // SAFETY:
+        // `tables` and `archetypes` belong to the same world that the [`QueryIter`]
+        // was initialized for.
+        unsafe {
+            location = self.entities.get(entity).debug_checked_unwrap();
+            archetype = self
+                .archetypes
+                .get(location.archetype_id)
+                .debug_checked_unwrap();
+            table = self.tables.get(location.table_id).debug_checked_unwrap();
+        }
+
+        // SAFETY: `archetype` is from the world that `fetch` was created for,
+        // `fetch_state` is the state that `fetch` was initialized with
+        unsafe {
+            D::set_archetype(
+                &mut self.fetch,
+                &self.query_state.fetch_state,
+                archetype,
+                table,
+            );
+        }
+
+        // The entity list has already been filtered by the query lens, so we forego filtering here.
+        // SAFETY:
+        // - set_archetype was called prior, `location.archetype_row` is an archetype index in range of the current archetype
+        // - fetch is only called once for each entity.
+        unsafe { D::fetch(&mut self.fetch, entity, location.table_row) }
+    }
+
+    /// Collects the internal [`I`](QuerySortedManyIter) once.
+    /// [`fetch_next`](QuerySortedManyIter) and [`fetch_next_back`](QuerySortedManyIter) require this to be called first.
+    #[inline(always)]
+    pub fn collect_inner(self) -> QuerySortedManyIter<'w, 's, D, F, IntoIter<Entity>> {
+        QuerySortedManyIter {
+            entity_iter: self.entity_iter.collect::<Vec<_>>().into_iter(),
+            entities: self.entities,
+            tables: self.tables,
+            archetypes: self.archetypes,
+            fetch: self.fetch,
+            query_state: self.query_state,
+        }
+    }
+}
+
+impl<'w, 's, D: QueryData, F: QueryFilter> QuerySortedManyIter<'w, 's, D, F, IntoIter<Entity>> {
+    /// Get next result from the query
+    /// [`collect_inner`](QuerySortedManyIter) needs to be called before this method becomes available.
+    /// This is done to prevent mutable aliasing.
```

Comment:
Let me rephrase for clarity:
In an iteration with no duplicates (`QuerySortedIter`) this presents no issue because the offending reference is always dropped before `D::Item` is retrieved.
However with duplicates, a mutable reference to `D::Item` could alias with a not yet dropped `L::Item` later on in the iteration, which means all `L::Item`s need to be dropped before we allow mutable iteration via `fetch_next`.

`ManuallyDrop` does not make sense here, since you either want to drop nothing in advance (lazy map) or everything (eager map).
The problem here is that iterators are lazy, and especially when sorting, you may just want the first or last item.
Paying for an eager map of the entire query seemed wasteful, so I implemented `collect_inner`.
This problem stems from mutable and immutable iteration returning the same struct. 

Maybe we could actually have two impl blocks for `QueryManyIter`, one for mutable `QueryData`, and one for immutable. It would be nice if you could have sort methods with the same name on both, returning an eagerly mapped or lazily mapped iterator respectively.

Reply:
Okay, it sounds like I understand the aliasing issue!  

Regarding eager vs. lazy iteration: Note that you still have to drop all of the `L::Item`s even if you just take the first or last item.  It just happens implicitly when you drop the `Map<IntoIter>` without fully iterating it.  

(And, of course, most *actual* `L::Item` types like `&T` and `Mut<T>` have no drop glue.  So we're doing nothing, and the question is *when* we're doing nothing :).)  

The expensive thing we want to avoid in the immutable case is having to allocate a fresh `Vec<Entity>` and copy everything over.  And it would be even better if we could avoid it in the mutable case, too!  

I think something like this works: 
```rust
let mut keyed_query: Vec<_> = query_lens
    .map(|(key, entity)| (ManuallyDrop::new(key), NeutralOrd(entity)))
    .collect();
keyed_query.sort();
for (item, _) in &mut keyed_query {
    // SAFETY: `keyed_query` is immediately consumed by the `map`, which ignores the item,
    // so the value is never exposed to safe code or dropped again.
    unsafe { ManuallyDrop::drop(item) };
}
let entity_iter = keyed_query.into_iter().map(|(.., entity)| entity.0);
```

We're eagerly dropping the lens items, but we have to drop those eventually, so that's moving work around rather than adding anything.  And for the common case where there is no drop glue, the compiler can eliminate the entire loop.  

There is no extra `collect()`, so there is no extra memory allocation or copying.  

And the lens items have all been dropped, so there are no longer aliasing concerns and we can offer `fetch_next` to mutable queries!

Reply:
I don't think this will compile/is safe.
See this example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=df1f5c8daf4f12206f4e954701efc3b7

Calling the destructor of a reference will not end its borrow, since that doesn't deallocate!

Reply:
It does compile!  But further testing with miri shows that it isn't sound :(.  I assumed that values wrapped in `ManuallyDrop` no longer need to be valid, but it seems that they do.   (But maybe [that will change?](https://github.com/rust-lang/rfcs/blob/master/text/3336-maybe-dangling.md))  

It does work if we use `MaybeUninit`, though!  We can drop its contents early, and rust doesn't require that they remain valid since they may be uninitialized.  It's more awkward to use, and doesn't `impl Ord`, but I got something that works and seems to make miri happy!  

I don't have permission to make a PR to this PR, so I'll push up the code as a PR in my own repo: chescock/bevy#6

Reply:
Hmm, I am still skeptical? The difference between `mem::drop` and `drop_in_place`/`Drop::drop` is that only `mem::drop` discards/deallocates the value. For the other two, a reference will still be valid after having been dropped! (It is Copy, so has no drop glue).
As such, the for loop with `MaybeUninit::assume_init_drop` does not actually do anything.
Whether this is sound or not I honestly don't know, and even if miri doesn't catch something doesn't mean it is sound.
If it were to be, how does the generated code change? By using a wrapper that opts out of validity like this, we might lose out on optimizations just as much vs if we had just `mem::drop`ped them all with a `collect`

Reply:
Note that in the playground, wrapping with `MaybeUninit` also fails: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=28215602287845f4679dc0f9bdabcbcd

Reply:
It doesn't seem like we'll get it figured out in a comment thread here, and it's tricky enough that it definitely needs more eyes on it, so I'll plan to push it up as a separate PR once this one is merged.  I'm hoping we agree that if we *can* find a way to do this soundly, that it would be a nice improvement to the API!  :)

Reply:
Feel free to open that PR after! If it can be done soundly and is better than just a plain collect, it'd be nice!